packages/flutter/test/cupertino/date_picker_test.dart

```diff
@@ -223,4 +193,238 @@ void main() {
+          child: new Directionality(
+            textDirection: TextDirection.ltr,
+            child: new CupertinoDatePicker(
+              mode: CupertinoDatePickerMode.date,
+              onDateTimeChanged: (_) {},
+              initialDateTime: new DateTime(2018, 1, 1, 10, 30),
+            ),
+          ),
+        ),
+      );
+
+      // Distance between the first and the last column should be the same.
+      expect(
+        tester.getCenter(find.text('January')).dx - tester.getCenter(find.text('2018')).dx,
+        distance,
+      );
+    });
+
+    testWidgets('width of picker in time mode is consistent', (WidgetTester tester) async {
+      await tester.pumpWidget(
+        new SizedBox(
+          height: 400.0,
+          width: 400.0,
+          child: new Directionality(
+            textDirection: TextDirection.ltr,
+            child: new CupertinoDatePicker(
+              mode: CupertinoDatePickerMode.time,
+              onDateTimeChanged: (_) {},
+              initialDateTime: new DateTime(2018, 1, 1, 10, 30),
+            ),
+          ),
+        ),
+      );
+
+      // Distance between the first column and the last column.
+      final double distance =
+          tester.getCenter(find.text('10')).dx - tester.getCenter(find.text('AM')).dx;
+
+      await tester.pumpWidget(
+        new SizedBox(
+          height: 400.0,
+          width: 800.0,
+          child: new Directionality(
+            textDirection: TextDirection.ltr,
+            child: new CupertinoDatePicker(
+              mode: CupertinoDatePickerMode.time,
+              onDateTimeChanged: (_) {},
+              initialDateTime: new DateTime(2018, 1, 1, 10, 30),
+            ),
+          ),
+        ),
+      );
+
+      // Distance between the first and the last column should be the same.
+      expect(
+        tester.getCenter(find.text('10')).dx - tester.getCenter(find.text('AM')).dx,
+        distance,
+      );
+    });
+
+    testWidgets('picker automatically scrolls away from invalid date', (WidgetTester tester) async {
+      DateTime date;
+      await tester.pumpWidget(
+        new SizedBox(
+          height: 400.0,
+          width: 400.0,
+          child: new Directionality(
+            textDirection: TextDirection.ltr,
+            child: new CupertinoDatePicker(
+              mode: CupertinoDatePickerMode.date,
+              onDateTimeChanged: (DateTime newDate) {
+                date = newDate;
+              },
+              initialDateTime: new DateTime(2018, 3, 30),
+            ),
+          ),
+        ),
+      );
+
+      await tester.drag(find.text('March'), const Offset(0.0, 32.0));
+
+      // This test currently fails because of an issue with FixedExtentScrollController.animateToItem().
+//      expect(
+//        date,
+//        new DateTime(2018, 2, 28),
+//      );
+    });
+
+    testWidgets('picker automatically scrolls the am/pm column when the hour column changes enough', (WidgetTester tester) async {
```

Comment:
Add a test for the other case you mentioned where you scroll from, say 10 PM to 3 AM or vice versa.